frame/support/src/traits.rs

```diff
@@ -1664,6 +1666,66 @@ pub trait Instance: 'static {
     const PREFIX: &'static str ;
 }
 
+/// A trait similar to `Convert` to convert values from `B` an abstract balance type
+/// into u64 and back from u128. (This conversion is used in election and other places where complex
+/// calculation over balance type is needed)
+///
+/// Total issuance of the currency is passed in, but an implementation of this trait may or may not use it.
+pub trait CurrencyToVote<B> {
+	/// Convert balance to u64.
+	fn to_vote(value: B, issuance: B) -> u64;
+
+	/// Convert u128 to balance.
+	fn to_currency(value: u128, issuance: B) -> B;
+}
+
+/// An implementation of `CurrencyToVote` tailored for chain's that have a balance type of u128.
+///
+/// The factor is the `(total_issuance / u64::max()).max(1)`, represented as u64. Let's look at the
+/// important cases:
+///
+/// If the chain's total issuance is less than u64::max(), this will always be 1, which means that
+/// the factor will not have any effect. In this case, any account's balance is also less. Thus,
+/// both of the conversions are basically an `as`; Any balance can fit in u64.
+///
+/// If the chain's total issuance is more than 2*u64::max(), then a factor might be multiplied and
+/// divided upon conversion.
+pub struct U128CurrencyToVote;
+
+impl U128CurrencyToVote {
+	fn factor(issuance: u128) -> u128 {
+		(issuance / u64::max_value() as u128).max(1)
+	}
+}
+
+impl CurrencyToVote<u128> for U128CurrencyToVote {
+	fn to_vote(value: u128, issuance: u128) -> u64 {
```

Comment:
The result of this function will look like a sawtooth as we increase the value of some user which increases the issuance past some threshold in `factor`.

i.e. when `issuance` increases such that `factor` jumps from 2 to 3, then `to_vote` for all the same values will be a fraction smaller than they were before.

Is this okay? Is it possible we are doing some calculation where some users are measured under one `factor` and other users are measured under another?

i.e. we are storing the vote value somewhere?

Reply:
> Is this okay? Is it possible we are doing some calculation where some users are measured under one factor and other users are measured under another?

No, because of what is discussed here: https://github.com/paritytech/substrate/pull/6896#discussion_r472053563 So as long as total_issuance stays the same, this function is also the same. 

> i.e. we are storing the vote value somewhere?

We don't store anything in vote count anywhere. In that case it would be a problem as total_issuance can change. 

> i.e. when issuance increases such that factor jumps from 2 to 3, then to_vote for all the same values will be a fraction smaller than they were before.


This is originally @gavofyork's code and I think the logic goes like this: If total issuance is below u64::max, then any account's balance is also within the range so we do `vote = balance / 1`. If total issuance is x times more than an account's balance can also be at most x times more than u64::max, thus `vote = balance / x`. And of course this is not 100% accurate (i.e. if total issuance is u64::max * 3 / 2), but as the documentation says this type must be _sensibly_ lossy.

Reply:
Maybe some comment about this could be added to the CurrencyToVote trait, indeed vote are not supposed to be stored but just used as intermediary arithmetic type for getting some result

Reply:
big agree with @thiolliere

Reply:
added.